File: app/src/main/java/com/ada/ada_meethem/ui/HomeFragment.java
package com.ada.ada_meethem.ui;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.ada.ada_meethem.R;
import com.ada.ada_meethem.adapters.PlanListAdapter;
import com.ada.ada_meethem.modelo.Group;
import com.ada.ada_meethem.modelo.Person;
import com.ada.ada_meethem.modelo.Plan;

import java.util.ArrayList;
import java.util.List;

public class HomeFragment extends Fragment {

    private List<Plan> plans;

    public HomeFragment() {
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View root = inflater.inflate(R.layout.fragment_home, container, false);

        generatePlans();

        RecyclerView plansRecyclerView = root.findViewById(R.id.plansRecyclerView);
        plansRecyclerView.setHasFixedSize(true);

        RecyclerView.LayoutManager layoutManager = new LinearLayoutManager(root.getContext());
        plansRecyclerView.setLayoutManager(layoutManager);

        PlanListAdapter plAdapter= new PlanListAdapter(plans,
                plan -> {
                });
        plansRecyclerView.setAdapter(plAdapter);

        return root;
    }

    private void generatePlans() {
        Person person1 = new Person("Maricarmen");
        Person person2 = new Person("Alex");
        Person person3 = new Person("Abel");
        Person person4 = new Person("Diego");

        Group group1 = new Group("Montañeros raperos", person1);
        Group group2 = new Group("Molamos", person2);

        Plan plan1 = new Plan("Ruta del Cares", group1, person1, 10, "https://www.escapadarural.com/blog/wp-content/uploads/2019/11/ruta-cares.jpg");
        Plan plan2 = new Plan("Padelcito", group2, person3, 3, "https://www.covb.cat/wp-content/uploads/2023/04/padel-g9a1278dbc_1280.jpg");
        Plan plan3 = new Plan("Surf en rodiles", group2, person4, 5, "https://img.redbull.com/images/q_auto,f_auto/redbullcom/2015/01/22/1331701014293_2/iv%C3%A1n-villalba-en-rodiles-%28asturias%29-semeyadetoral.com.jpg");

        plan3.addToPlan(person1);
        plan3.addToPlan(person3);
        plan2.addToPlan(person1);
        plan2.addToPlan(person2);
        plan2.addToPlan(person4);

        plans = new ArrayList<>();
        plans.add(plan1);
        plans.add(plan2);
        plans.add(plan3);
        plans.add(plan1);
        plans.add(plan2);
        plans.add(plan3);
        plans.add(plan1);
        plans.add(plan2);
        plans.add(plan3);
        plans.add(plan1);
        plans.add(plan2);
        plans.add(plan3);
    }

}

File: app/src/main/java/com/ada/ada_meethem/ui/HomeFragment.kt
package com.ada.ada_meethem.ui

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ProgressBar
import android.widget.TextView
import androidx.activity.result.contract.ActivityResultContracts
import androidx.fragment.app.Fragment
import androidx.lifecycle.lifecycleScope
import androidx.navigation.fragment.findNavController
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.ada.ada_meethem.R
import com.ada.ada_meethem.adapters.PlanListAdapter
import com.ada.ada_meethem.database.ContactDatabase
import com.ada.ada_meethem.database.PlanDatabase
import com.ada.ada_meethem.database.entities.Contact
import com.ada.ada_meethem.modelo.Plan
import com.ada.ada_meethem.ui.PlanFragment.Companion.newInstance
import com.ada.ada_meethem.util.ContactLoaderFromProvider
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class HomeFragment : Fragment() {

    private var plans: List<Plan> = ArrayList()
    private var contacts: List<Contact> = ArrayList()
    private lateinit var plAdapter: PlanListAdapter
    private lateinit var contactLoader: ContactLoaderFromProvider
    private lateinit var progressBar: ProgressBar
    private lateinit var progressBarText: TextView

    private var loadContacts: Boolean = true

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setUsersListener()

        // Crea el contactLoader
        contactLoader = ContactLoaderFromProvider(
            requireContext(),
            requestPermissionLauncher,
            requireActivity()
        )
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        // Inflate the layout for this fragment
        val root = inflater.inflate(R.layout.fragment_home, container, false)

        progressBar = root.findViewById(R.id.progressBar)
        progressBarText = root.findViewById(R.id.progressBarText)

        lifecycleScope.launch(Dispatchers.IO) {
            withContext(Dispatchers.Main) {
                progressBar.visibility = View.VISIBLE
                progressBarText.visibility = View.VISIBLE
            }

            // Carga los contactos y los planes
            if (loadContacts) {
                contacts = contactLoader.loadContacts()
                loadContacts = false
            }
            loadPlans()

            withContext(Dispatchers.Main) {
                progressBar.visibility = View.GONE
                progressBarText.visibility = View.GONE

            }
        }

        // Configura el recycler view de planes
        val plansRecyclerView = root.findViewById<RecyclerView>(R.id.plansRecyclerView)
        plansRecyclerView.setHasFixedSize(true)
        val layoutManager: RecyclerView.LayoutManager = LinearLayoutManager(root.context)
        plansRecyclerView.layoutManager = layoutManager
        plAdapter = PlanListAdapter(plans) { plan -> clickonItem(plan) }
        plansRecyclerView.adapter = plAdapter
        return root
    }

    fun clickonItem(plan: Plan?) {
        //Paso el modo de apertura
        val planFragment = newInstance()
        val bundle = Bundle()
        bundle.putParcelable("plan", plan)
        //planFragment.setArguments(bundle);
        //getFragmentManager().beginTransaction().replace(R.id.fragment_container, planFragment).commit();
        this.findNavController().navigate(
            R.id.action_homeFragment_to_planFragment,
            bundle
        )
    }

    private fun loadPlans() {
        val phoneNumber = FirebaseAuth.getInstance().currentUser!!.phoneNumber
        PlanDatabase.getReference().addValueEventListener(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                val plans = ArrayList<Plan>()
                val cdb = ContactDatabase.getDatabase(context).contactDAO
                for (chatSnapshot in dataSnapshot.children) {
                    val plan = chatSnapshot.getValue(Plan::class.java) as Plan
                    if (plan.enlisted.contains(phoneNumber)) {
                        plans.add(plan)
                    }
                }
                plAdapter.update(plans)
            }

            override fun onCancelled(databaseError: DatabaseError) {
                // Manejo de errores
            }
        })
    }

    private fun setUsersListener() {
        val databaseReference =
            FirebaseDatabase.getInstance("https://meethem-8955a-default-rtdb.europe-west1.firebasedatabase.app/")
                .getReference("users").addValueEventListener(object : ValueEventListener {
                    override fun onDataChange(dataSnapshot: DataSnapshot) {
                        val cdb = ContactDatabase.getDatabase(context).contactDAO
                        for (chatSnapshot in dataSnapshot.children) {
                            val contact = chatSnapshot.getValue(Contact::class.java) as Contact
                            val localContact = cdb.findByNumber(contact.contactNumber)
                            if (localContact != null) {
                                localContact.photoUrl = contact.photoUrl
                            }
                        }
                    }

                    override fun onCancelled(databaseError: DatabaseError) {
                        // Manejo de errores
                    }
                })

    }


    private val requestPermissionLauncher = registerForActivityResult(
        ActivityResultContracts.RequestPermission()
    )
    { isGranted: Boolean -> contactLoader.requestPermissionLauncherCallback(isGranted) }

}